### PR TITLE
Complete Assignment_1

### PR DESCRIPTION
**Pull request by meisammofidi**
## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)
Adding code

## What did you learn from the changes you have made?
Data modeling

## Checklist
- [ ] I can confirm that my changes are working as intended

### CODE REVIEW

**Comment by efantinatti** [comment]
# PR Feedback Report

**Student**: Meisam  
**GitHub Username**: meisammofidi  
**PR Number**: 1  
**Branch**: assignment_1  
**Status**: INCOMPLETE  
**PR URL**: https://github.com/meisammofidi/lcr/pull/1

## What You Did Well (Hits)

The student successfully loaded the dataset and performed initial data inspection, correctly identifying 178 observations and 14 variables. The response variable type was correctly identified as int64. The student also correctly implemented data standardization using StandardScaler and performed a train/test split. The grid search for KNeighborsClassifier was set up with the correct parameter range (1-50) and 10-fold cross-validation. The model evaluation on the test data was performed using accuracy_score.

## Areas for Improvement (Errors)

Answer for the levels of the response variable in Question 1 is missing. The best n_neighbors found by the grid search was 20, which does not match the expected value of 8. Consequently, the model evaluation in Question 4 resulted in an accuracy of 0.733, which is significantly lower than the expected 0.947.  Missing some outputs as well.

## Advice and Recommendations

For Question 1, ensure all parts of the question are answered, including the levels of the response variable. For Question 3, review the grid search setup and consider if any steps might lead to a different 'best n_neighbors' value. Double-check the random seed used in Question 2 to ensure it matches the specified value. It is important to ensure that the code produces the expected outputs as per the assignment requirements.

## Overall Assessment

Your submission demonstrates a good understanding of data inspection, standardization, and model evaluation. However, there are critical discrepancies in the results for best hyperparameter selection and final model accuracy, indicating areas for improvement in the implementation of the grid search and potentially the data splitting process.

**Comment by efantinatti** [comment]
Hi @meisammofidi can you please explain the changes?

**Comment by meisammofidi** [comment]
Hi @efantinatti 
Changes are availble in https://github.com/meisammofidi/LCR/pull/1/commits/dbb728c674b9568988a52e8b4db73af508412575

# Summary:

_(iii)_ What is the 'variable type' of the response variable `class` (e.g., 'integer', 'category', etc.)? What are the 'levels' (unique values) of the variable?
wine_df['class'].info()  # Data type of 'class' column
print(wine_df['class'].unique())  # Unique values in 'class' column --> +
print(wine_df['class'].value_counts())  # Count of each unique value --> +
print(wine_df['class'].describe())  # Summary statistics --> +


# Question 3: I have updated code as follows. Still return 20 as best parameter for K

X_train = wine_training.drop(columns=['class'])
y_train = wine_training['class']

# 1) Initialize KNN
knn = KNeighborsClassifier()

# 2) Parameter grid: n_neighbors = 1..50
parameter_grid = {
    "n_neighbors": range(1, 50),
}

# 3) Grid search with 10-fold CV (Stratified by default for classifiers)
wine_grid = GridSearchCV(
    estimator=knn,
    param_grid=parameter_grid,
    cv=10,
)

# Fit on training data
wine_grid.fit(X_train, y_train)

# 4) Best k
wine_grid.best_params_

{'n_neighbors': 20}

================

# Question 4:

# Use the best k from GridSearchCV
best_k = wine_grid.best_params_["n_neighbors"]

# Initialize KNN with best k
best_knn = KNeighborsClassifier(n_neighbors=best_k)

# Fit on training data
best_knn.fit(X_train, y_train)
best_knn

# Predict on test data
wine_testing['predicted'] = best_knn.predict(wine_testing.drop(columns='class'))
wine_testing[['predicted', 'class']]

**Comment by efantinatti** [comment]
Hi @meisammofidi looks like the scaled data is not being used for test and train split. This explains the low accuracy of the model.

**Comment by efantinatti** [comment]
Hi @meisammofidi congratulations, you standardized your train and test and now your result is great!

COMPLETED!
